lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp

```diff
@@ -1513,30 +1513,96 @@ chooseMfmaLikeStoreLayout(RankedTensorType valType) {
 
   // We currently only support transposed [B]F16 MFMA32x32 on CDNA4.
   bool isMfma32 = mfmaLayout.getMDim() == 32 && mfmaLayout.getNDim() == 32;
+  bool isMfma16 = mfmaLayout.getMDim() == 16 && mfmaLayout.getNDim() == 16;
+
+  auto valShape = valType.getShape();
+  // For mfma16x16, to use in-wave swap, we need to make sure the tiles used are
+  // in one wave if there are multiple tiles.
+  bool validForMfma16 = isMfma16 && valShape.back() >= 16 * 2 &&
+                        mfmaLayout.getWarpsPerCTA().back() == 1;
+
   Type elemType = valType.getElementType();
   if (!(valType.getRank() == 2 && (elemType.isF16() || elemType.isBF16()) &&
         mfmaLayout.getVersionMajor() == 4 && mfmaLayout.getIsTransposed() &&
-        isMfma32))
+        (isMfma32 || validForMfma16)))
     return {};
 
-  auto valShape = valType.getShape();
   LinearLayout mfmaLL = mfmaLayout.toLinearLayout(valShape);
   auto mfmaOutDims = llvm::to_vector(mfmaLL.getOutDimNames());
   StringAttr dimM = mfmaOutDims[0];
   StringAttr dimN = mfmaOutDims[1];
-
   auto swapLL = LinearLayout::empty();
   // The rows are kept as is with an identity linear layout.
   swapLL *= LinearLayout::identity1D(valShape[0], dimM, dimM);
-  // In transposed mfma32 layout, each thread holds 4 consecutive values along N
-  // dim. We want to exchange column 4-7 (owned by thread 32-63) and column 8-11
-  // (owned by thread 0-31) every 16 columns to make each thread holds 8
-  // elements. This would mean exchange the 2nd and 3rd basis vector from an
-  // identity linear layout.
+  /*
+      In transposed mfma32 layout,
+      1) register is the N or column dimension and lane is the M or row
+  dimension;
+
+      2) the pair, e.g.(0, 0) is for the indices in the tensor;
+            register/N
+  lane/M    (0,  0) ... (0,  3)  | (0,  8)  ... (0, 11) | ...
+            ...                  | BLK1                 |
+            (31, 0) ... (31, 3)  | (31, 8)  ... (31, 11)| ...
+            .................................................
+          |  (0,  4) ... (0,  7) |  (0,  12) ... (0, 15)  ...
+          |  BLK0                |
+          |  (31, 4) ... (31, 7) |  (31, 12) ... (31, 15) ...
+          ........................
+      each thread holds 4 consecutive values along N dim.
+      We want to exchange column 4-7 (owned by thread 32-63, BLK0) and column
+    8-11 (owned by thread 0-31, BLK1) every 16 columns to make each thread holds
+    8 elements. This would mean exchange the 2nd and 3rd basis vector from an
+      identity linear layout on tensor elements.
+      Correspondingly, the transposed mfma16 layout,
+      the output of mfma16x16: transposed
```

Comment:
Can you also see weird indentation of this comment block? If so, can you fix it?